src/core/math/matrixZipMap.ts

```diff
@@ -2,6 +2,30 @@ import * as R from 'ramda';
 
 type Matrix<T> = T[][];
 
+export const shallowClone: <T>(
+    m: Matrix<T>
+) => Matrix<T> = R.map(row => row.slice(0));
+
+export const traverse2 = <T1, T2, TR>(
+    a1: T1[],
+    a2: T2[],
+    fn: (d1: T1, d2: T2, i1: number, i2: number) => TR
+) => R.addIndex<T1>(R.forEach)((d1, i1) =>
+    R.addIndex<T2>(R.forEach)((d2, i2) =>
+        fn(d1, d2, i1, i2),
+        a2),
+    a1);
```

Comment:
Utility. Apply `fn` for all combinations `(a,b)` of lists `A, B`

Reply:
I have to wonder whether all these little utilities wouldn't be better off without `ramda` at all - in plotly.js we don't even use `Array.forEach` for really hot paths, only `for`:
```js
for(let i1 = 0; i1 < a1.length; i1++) {
    for(let i2 = 0; i2 < a2.length; i2++) {
        fn(a1[i1], a2[i2], i1, i2);
    }
}
```

Reply:
Based on this comment, removed Ramda from the utility functions to see what sort of impact it has.

On average, through 7 renders (don't ask.. glad you did!.. my performance tab was opened in such a way that ~7 renders would fit..) seeing a small but consistent 10-15% speed increase for the utility calls with the changes in https://github.com/plotly/dash-table/pull/465/commits/62e519d2bee18b6a305ab6438cb3ee3935c4581b. Real world gains are in the order of 3-4ms per render cycle.. or about 2% of the render time.

This code is very isolated/abstracted so I see no impact to adding these changes in terms maintainability as long as all the tests are still passing. To be confirmed!